### PR TITLE
`ultralytics 8.2.30` automated tags and release notes

### PR DESCRIPTION
**Pull request by IvorZhu331**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to the GitHub Actions workflow for automated version publishing and documentation updates, plus minor documentation improvements.

### 📊 Key Changes
- Introduced automated Git configuration for the publishing workflow.
- Automated the creation and publishing of new version tags and releases based on version increments.
- Added an AI-based summary generation for release notes using OpenAI.
- Minor whitespace adjustments across various documentation files.
- Updated version from 8.2.29 to 8.2.30.

### 🎯 Purpose & Impact
- **Automation**: Streamlines the release process by automating version tagging and release creation when updates are detected.
- **Consistency**: Ensures consistent Git user configuration in the publishing workflow.
- **Documentation Quality**: Improved readability with minor punctuation and spacing fixes.
- **Efficiency**: Saves time for developers by automating repetitive tasks and providing clear, concise release summaries.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@IvorZhu331 we need to transition this PR over to our Azure keys and endpoints just like we do for Ultralytics Actions here:
https://github.com/ultralytics/ultralytics/blob/bddea17bf390cfc8b37b93b35133b6398db4fca1/.github/workflows/format.yml#L28-L29

Please work with @fcakyon to get this transitioned using the same keys as above. Thank you!

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -56,10 +57,72 @@ jobs:
...
+          response = requests.get(compare_tags_url, headers=headers).json()
+          diffs = []
+          files = response.get('files', [])
+          for file in files:
+              patch = file.get('patch', '')
+              if patch:
+                  diffs.append(patch)       
+          combined_diff = '\n'.join(diffs)
+          limit = 64000 * 3 # can be changed if needed
+          client = openai.OpenAI(api_key=OPENAI_API_KEY)
```
Comment (fcakyon) [inline .github/workflows/publish.yml:100]:
```suggestion
          client = openai.AzureOpenAI(api_key=OPENAI_AZURE_API_KEY, api_version=OPENAI_AZURE_API_VERSION, azure_endpoint=OPENAI_AZURE_ENDPOINT)
```

Comment (fcakyon) [reply .github/workflows/publish.yml:100]: where OPENAI_AZURE_ENDPOINT = "2024-02-01"

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -56,10 +57,72 @@ jobs:
...
+              {"role": "user", "content": f"Summarize the updates made on '{latest_tag}' tag, focusing on major changes, their purpose, and potential impact. Keep the summary clear and concise, suitable for a broad audience. Add emojis to enliven the summary. Reply directly with a summary along these example guidelines, though feel free to adjust as appropriate:\n\n"
+              f"### 🌟 Summary (single-line synopsis)\n"
+              f"### 📊 Key Changes (bullet points highlighting any major changes)\n"
+              f"### 🎯 Purpose & Impact (bullet points explaining any benefits and potential impact to users)\n"
+              f"\n\nHere's the PR diff:\n\n{combined_diff[:limit]}",
+              }
+            ]
+          response = client.chat.completions.create(model="gpt-4o", messages=messages).choices[0]
```
Comment (fcakyon) [inline .github/workflows/publish.yml:110]:
```suggestion
          response = client.chat.completions.create(model="gpt-4-turbo-2024-04-09", messages=messages).choices[0]
```

File: .github/workflows/publish.yml (hunk 1 above)
Comment (fcakyon) [reply .github/workflows/publish.yml:100]: @IvorZhu331 sent the other info from slack 👍🏻

**Comment by fcakyon** [comment]
> @IvorZhu331 we need to transition this PR over to our Azure keys and endpoints just like we do for Ultralytics Actions here:
> 
> https://github.com/ultralytics/ultralytics/blob/bddea17bf390cfc8b37b93b35133b6398db4fca1/.github/workflows/format.yml#L28-L29
> 
> Please work with @fcakyon to get this transitioned using the same keys as above. Thank you!

Hey @glenn-jocher, used model in https://github.com/ultralytics/ultralytics/blob/main/.github/workflows/format.yml seems to be outdated. Should we update it to gpt4-o as well?

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -56,10 +57,74 @@ jobs:
...
           os.system(f'echo "version={ultralytics.__version__}" >> $GITHUB_OUTPUT')
+          os.system(f'echo "previous_version={check_latest_pypi_version()}" >> $GITHUB_OUTPUT')
 
           if increment:
               print('Local version is higher than PyPI version. Publishing new version to PyPI ✅.')
         id: check_pypi
+      - name: Summarize version changes
+        env:
+          OPENAI_AZURE_API_KEY: ${{ secrets.OPENAI_AZURE_API_KEY }} 
+          OPENAI_AZURE_ENDPOINT: ${{ secrets.OPENAI_AZURE_ENDPOINT }}
```
Comment (fcakyon) [inline .github/workflows/publish.yml:68]: I need to find out which region these secrets refer to. Do you know who set them @glenn-jocher ?

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -56,10 +57,74 @@ jobs:
...
+              {"role": "user", "content": f"Summarize the updates made on '{latest_tag}' tag, focusing on major changes, their purpose, and potential impact. Keep the summary clear and concise, suitable for a broad audience. Add emojis to enliven the summary. Reply directly with a summary along these example guidelines, though feel free to adjust as appropriate:\n\n"
+              f"### 🌟 Summary (single-line synopsis)\n"
+              f"### 📊 Key Changes (bullet points highlighting any major changes)\n"
+              f"### 🎯 Purpose & Impact (bullet points explaining any benefits and potential impact to users)\n"
+              f"\n\nHere's the PR diff:\n\n{combined_diff[:limit]}",
+              }
+            ]
+          response = client.chat.completions.create(model="gpt-4-turbo-2024-04-09", messages=messages).choices[0]
```
Comment (fcakyon) [inline .github/workflows/publish.yml:112]: Can you try with gpt-4-0125-preview @IvorZhu331 ?

**Review by fcakyon — CHANGES_REQUESTED** [review]

File: .github/workflows/publish.yml (hunk 4 above)
Comment (IvorZhu331) [reply .github/workflows/publish.yml]: Azure API is now working with `gpt-4-0125-preview`.

**Review by glenn-jocher — APPROVED** [review]

**Pull request merged**